diff -rpuN ns-3.21-backup/src/netanim/model/animation-interface.cc ns-3.21/src/netanim/model/animation-interface.cc
--- ns-3.21-backup/src/netanim/model/animation-interface.cc	2014-12-02 20:51:57.485842717 +0800
+++ ns-3.21/src/netanim/model/animation-interface.cc	2014-12-06 23:07:01.542555957 +0800
@@ -1514,7 +1514,8 @@ AnimationInterface::GetPosition (Ptr <No
 {
   if (m_nodeLocation.find (n->GetId ()) == m_nodeLocation.end ())
     {
-      NS_FATAL_ERROR ("Node:" <<n->GetId () << " not found in Location table");
+      //NS_FATAL_ERROR ("Node:" <<n->GetId () << " not found in Location table");
+      m_nodeLocation[n->GetId ()] = Vector (0, 0, 0);
     }  
   return m_nodeLocation[n->GetId ()];
 }
@@ -1579,7 +1580,10 @@ AnimationInterface::WriteLinkProperties
             }
           std::string channelType = ch->GetInstanceTypeId ().GetName ();
           NS_LOG_DEBUG ("Got ChannelType" << channelType);
-          if (channelType == std::string ("ns3::PointToPointChannel"))
+          if (channelType == std::string ("ns3::PointToPointChannel") ||
+              channelType == std::string ("ns3::CsmaChannel") ||
+              channelType == std::string ("ns3::SimpleChannel"))
+
             { // Since these are duplex links, we only need to dump
               // if srcid < dstid
               uint32_t nChDev = ch->GetNDevices ();
